pkgs/build-support/rust/rust-utils.nix

```diff
@@ -0,0 +1,270 @@
+# Copyright 2017 Pierre-Étienne Meunier
+#
+# This file is licensed under the Apache-2.0 license, and under the
+# MIT license, at your convenience.
+
+{ lib, buildPlatform, stdenv }:
+
+let buildCrate = { crateName, crateVersion, dependencies, complete, crateFeatures, libName, build, release, libPath, crateType, metadata, crateBin, finalBins, verboseBuild }:
+
+      let depsDir = builtins.foldl' (deps: dep: deps + " " + dep.out) "" dependencies;
+          completeDepsDir = builtins.foldl' (deps: dep: deps + " " + dep.out) "" complete;
+          deps =
+            builtins.foldl' (deps: dep:
+              let extern = lib.strings.replaceStrings ["-"] ["_"] dep.libName; in
+              deps + (if dep.crateType == "lib" then
+                 " --extern ${extern}=${dep.out}/lib${extern}-${dep.metadata}.rlib"
+              else
+                 " --extern ${extern}=${dep.out}/lib${extern}-${dep.metadata}.so")
+            ) "" dependencies;
```

Comment:
`foldl'` can be here also replaced by `concatMapStringsSep " " (dep: ...) dependencies`